urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/SSLFactory.java

```diff
@@ -0,0 +1,149 @@
+package org.javastack.bouncer;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.Socket;
+import java.security.GeneralSecurityException;
+import java.security.KeyFactory;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.spec.PKCS8EncodedKeySpec;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLServerSocket;
+import javax.net.ssl.SSLServerSocketFactory;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManagerFactory;
+import javax.xml.bind.DatatypeConverter;
+
+public class SSLFactory {
+	private final static char[] DEFAULT_PWD = "changeit".toCharArray();
+	private final KeyStore ks;
+	private final SSLContext ctx;
+	private final CipherSuites cipherSuites;
+	private final boolean needClientCert;
+
+	public SSLFactory(final CipherSuites cipherSuites, final String priCert, final String priKey)
+			throws IOException, GeneralSecurityException {
+		this(cipherSuites, priCert, priKey, null);
+	}
+
+	public SSLFactory(final CipherSuites cipherSuites, final String priCert, final String priKey,
+			final String pubCert) throws IOException, GeneralSecurityException {
+		this.cipherSuites = cipherSuites;
+		this.ks = initKeyStore(loadX509(priCert), loadPriKey(priKey), loadX509(pubCert));
+		this.ctx = initSSLContext(ks);
+		this.needClientCert = ((pubCert != null) && !pubCert.isEmpty());
+	}
+
+	public SSLServerSocket createSSLServerSocket() throws IOException {
+		SSLServerSocketFactory factory = ctx.getServerSocketFactory();
+		SSLServerSocket listen = (SSLServerSocket) factory.createServerSocket();
+		listen.setEnabledProtocols(cipherSuites.getProtocols());
+		listen.setEnabledCipherSuites(cipherSuites.getServerCipherSuites());
+		listen.setNeedClientAuth(needClientCert); // Force Request Client Certificate
+		return listen;
+	}
+
+	public SSLSocket createSSLSocket() throws IOException {
+		SSLSocketFactory factory = ctx.getSocketFactory();
+		SSLSocket sock = (SSLSocket) factory.createSocket();
+		sock.setEnabledProtocols(cipherSuites.getProtocols());
+		sock.setEnabledCipherSuites(cipherSuites.getClientCipherSuites());
+		return sock;
+	}
+
+	public CipherSuites getCipherSuites() {
+		return cipherSuites;
+	}
+
+	public static PrivateKey loadPriKey(final String fileName) throws IOException, GeneralSecurityException {
+		PrivateKey key = null;
+		InputStream is = null;
+		try {
+			is = fileName.getClass().getResourceAsStream("/" + fileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;fileName&#39; (index: 0 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;loadX509&#39;. The same parameter type is validated here for example: urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/Bouncer.java:248. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).